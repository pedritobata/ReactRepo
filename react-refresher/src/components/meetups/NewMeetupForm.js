import { useRef } from "react";
import classes from "./NewMeetupForm.module.css";

const NewMeetupForm = (props) => {
  const titleRef = useRef();
  const imageRef = useRef();
  const addressRef = useRef();
  const descriptionRef = useRef();

  const submitHandler = (event) => {
      event.preventDefault();
      const meetup = {
          title: titleRef.current.value,
          image: imageRef.current.value,
          address: addressRef.current.value,
          description: descriptionRef.current.value,
      }
      props.onSubmitMeetup(meetup);
  };

  return (
    <form className={classes.form} onSubmit={submitHandler}>
      <div className={classes.control}>
        <label htmlFor="title">Title</label>
        <input type="text" id="title" name="title" required ref={titleRef} />
      </div>
      <div className={classes.control}>
        <label htmlFor="image">Image</label>
        <input type="url" id="image" name="image" ref={imageRef} />
      </div>
      <div className={classes.control}>
        <label htmlFor="address">Address</label>
        <input type="text" id="address" name="address" ref={addressRef} />
      </div>
      <div className={classes.control}>
        <label htmlFor="description">Description</label>
        <textarea
          id="description"
          name="description"
          cols={5}
          required
          ref={descriptionRef}
        ></textarea>
      </div>
      <div className={classes.actions}>
        <button>Add Meetup</button>
      </div>
    </form>
  );
};

export default NewMeetupForm;
