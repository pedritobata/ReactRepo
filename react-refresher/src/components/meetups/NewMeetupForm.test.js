import { expect, it } from '@jest/globals';
import { getDefaultNormalizer, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import NewMeetupForm from './NewMeetupForm';


describe('NewMeetupForm', () => {

    it('should render correctly', () => {
        const snap = renderer.create(<NewMeetupForm />).toJSON();
        expect(snap).toMatchSnapshot();
    })

    it('should render a title input field', () => {
        render(<NewMeetupForm />);
        screen.getByLabelText('Title');
    })

    it('should render an image field', () => {
        render(<NewMeetupForm />);
        screen.getByLabelText('Image');
    })

    it('should render an address field', () => {
        render(<NewMeetupForm />);
        screen.getByLabelText('Address', {exact: false})
    })

    it('should render a description textarea as the last field', () => {
        render(<NewMeetupForm />);
        screen.getByLabelText('Description', {
            exact: true,
            normalizer: (str) => getDefaultNormalizer({trim:  true})(str)
        });
        const fields = screen.getAllByRole('textbox');
        const descriptionField = fields[fields.length - 1];
        expect(descriptionField.name).toBe('description');
        expect(descriptionField.tagName).toMatch(/textarea/i);
    })

    it('should call parent onSubmitMeetup function when submitting', () => {
        const mockOnSubmitMeetup = jest.fn()
        render(<NewMeetupForm  onSubmitMeetup={mockOnSubmitMeetup} />);
        const meetup = {
            title: 'Perico los palotes',
            image: 'hhtp://la.imagen.jpg',
            address: 'Av Trinidad Moran',
            description: 'Esta es como la p...',
        }
        const title = screen.getByLabelText('Title');
        userEvent.type(title, meetup.title);
        const image = screen.getByLabelText('Image');
        userEvent.type(image, meetup.image);
        const address = screen.getByLabelText('Address');
        userEvent.type(address, meetup.address);
        const description = screen.getByLabelText('Description');
        userEvent.type(description, meetup.description);

        userEvent.click(screen.getByRole('button'));

        expect(mockOnSubmitMeetup).toHaveBeenCalledWith(meetup);
        
    })
})