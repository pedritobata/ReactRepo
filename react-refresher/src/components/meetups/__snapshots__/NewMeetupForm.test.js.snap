// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewMeetupForm should render correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="control"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <input
      id="title"
      name="title"
      required={true}
      type="text"
    />
  </div>
  <div
    className="control"
  >
    <label
      htmlFor="image"
    >
      Image
    </label>
    <input
      id="image"
      name="image"
      type="url"
    />
  </div>
  <div
    className="control"
  >
    <label
      htmlFor="address"
    >
      Address
    </label>
    <input
      id="address"
      name="address"
      type="text"
    />
  </div>
  <div
    className="control"
  >
    <label
      htmlFor="description"
    >
      Description
    </label>
    <textarea
      cols={5}
      id="description"
      name="description"
      required={true}
    />
  </div>
  <div
    className="actions"
  >
    <button>
      Add Meetup
    </button>
  </div>
</form>
`;
